text/3556-re-export-symbols.md

```diff
@@ -0,0 +1,131 @@
+
+When linking a staticlib into a cdylib crate, there are some cases where it is useful to re-export symbols. There is extensive dicussion from [people wanting to do this](https://github.com/rust-lang/rfcs/issues/2771) (with a note that the Lang team discussed this in 2017!), as well as another issue from people attempting to [use `+whole-archive`](https://github.com/rust-lang/rust/issues/110624) to re-export _all_ symbols from a linked staticlib.
+
+However, the *only* supported way to re-export a symbol today is manually write wrappers around the staticlib symbols, re-passing the arguments down. This is:
+ - inconvenient - re-exporting as the same name requires putting the `extern` block and wrapper function in different modules due to name collision
+ - error-prone - the user has to maintain the wrapper function with the correct arguments
+ - sometimes impossible - variadic functions are [not stable yet](https://github.com/rust-lang/rust/issues/44930)
+
+This tightly-scoped RFC adds support for re-exporting explicitly listed symbols from a linked staticlib in an extern block to the final produced cdylib. It does *not* add 'wildcard' support to support re-exporting all symbols from a staticlib, nor does it cover other staticlib/cdylib combinations.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When producing a cdylib from a Rust crate and linking to a staticlib, adding `#[no_mangle]` and `pub` to an item in an extern block in the top level crate as follows:
+
+```
+#[no_mangle]
+pub fn foo();
+```
+
+will make the `foo` symbol from the static library (e.g. `libext.a` on Linux) visible in the final produced `libcrate.so`.
+
+If the crate is not being build as a cdylib the `#[no_mangle]` annotation will be ignored - this includes when the crate is a dependency of a cdylib crate.
+
+If `ext` is not _explicitly_ linked as a static library the result is unspecified - the symbol may or may not be visible depending on the behavior of the linker. A non-fatal warning will be emitted:
+
+```
+warning: `#[no_mangle]` with `pub` is only valid for re-export when the library is explicitly a `staticlib`
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This feature [works today](https://github.com/aidanhs/rust-re-export-lib/) (at least on Linux), with a single caveat - a warning from the compiler:
+
+```
+   Compiling x v0.1.0 (/tmp/tmp.Sig2rLZept/rust-re-export-lib)
+warning: `#[no_mangle]` has no effect on a foreign function
+ --> src/lib.rs:5:5
+  |
+5 |     #[no_mangle]
+  |     ^^^^^^^^^^^^ help: remove this attribute
+6 |     pub fn foo();
+  |     ------------- foreign function
+  |
+  = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
```

Comment:
It would be nice if this included some history about why this deprecation was made. I assume it's probably because people weren't aware that kr actually does something, but it would be nice to confirm that.

Reply:
Done.